lib/lists.nix

```diff
@@ -193,6 +193,38 @@ rec {
       if len < 2 then list
       else (sort strictLess pivot.left) ++  [ first ] ++  (sort strictLess pivot.right));
 
+  # Depth-First Search (DFS) for `list != []`.
+  #
+  # `before a b == true` means that `b` depends on `a` (there's an
+  # edge from `b` to `a`).
+  listDfs = before: list:
+   let
+     dfs' = us: visited: rest:
+       let
+         b = partition (x: before x us) rest;
+       in if length b.right == 0
+          # if nothing is before us
+          then { minimal = us; inherit visited rest; }
+          # else grab the first one before us and continue
+          else dfs' (head b.right)
+                    ([ us ] ++ visited)
+                    (tail b.right ++ b.wrong);
+   in dfs' (head list) [] (tail list);
+
+  # Sort a list based on a partial ordering using DFS. This
+  # implementation is O(N^2) in the best case, if your ordering is
+  # linear, use `sort` instead.
+  #
+  # `before a b == true` means that `b` should be after `a`
+  # in the result.
+  toposort = before: list:
```

Comment:
yes, in comment. majority of this functions have nice examples which help a lot when going through them.